File: src/features/ruler/centerlines.ts
// SPDX-FileCopyrightText: 2026 overpolish
// SPDX-License-Identifier: GPL-3.0-or-later

import { RulerComponentBox } from "./api";
import { boxRect, contains, PixelRect } from "./bounds-snap";
import { Axis } from "./gradient-field";
import { Bounds } from "./pixel-analysis";

/** Centres this close in device px read as collinear. */
const ALIGNMENT_SLACK = 1;
/** Below this a component is noise - antialiasing, a hairline, a stray dot. */
const MINIMUM_OBJECT_SIZE = 3;
/** More outlines than this stops reading as "the things in the box". */
const MAXIMUM_OBJECTS = 12;

export type Centerline = {
  accent: boolean;
  position: number;
};

/** A piece of content inside a measurement, in world coordinates. */
export type InnerObject = {
  alignedX: boolean;
  alignedY: boolean;
  bounds: Bounds;
};

export type Centerlines = {
  objects: readonly InnerObject[];
  x: Centerline;
  y: Centerline;
};

const centreOf = (bounds: Bounds) => ({
  x: bounds.x + bounds.width / 2,
  y: bounds.y + bounds.height / 2,
});

/** A measurement that snapped onto a component matches it edge-for-edge. */
const SELF_SLACK = 3;

const isSelf = (rect: PixelRect, box: PixelRect) =>
  Math.abs(box.x0 - rect.x0) <= SELF_SLACK &&
  Math.abs(box.y0 - rect.y0) <= SELF_SLACK &&
  Math.abs(box.x1 - rect.x1) <= SELF_SLACK &&
  Math.abs(box.y1 - rect.y1) <= SELF_SLACK;

/**
 * What constitutes ONE object: components whose boxes come within this many
 * device px of each other cluster together, so a multi-stroke icon (or the
 * words of a label) reads as a single object instead of its parts.
 */
const CLUSTER_GAP = 6;
/** Clustering input cap - keeps the merge loop bounded on text-heavy boxes. */
const MAXIMUM_PARTS = 128;

const near = (a: PixelRect, b: PixelRect) =>
  a.x0 - CLUSTER_GAP <= b.x1 &&
  b.x0 - CLUSTER_GAP <= a.x1 &&
  a.y0 - CLUSTER_GAP <= b.y1 &&
  b.y0 - CLUSTER_GAP <= a.y1;

const merge = (a: PixelRect, b: PixelRect): PixelRect => ({
  x0: Math.min(a.x0, b.x0),
  x1: Math.max(a.x1, b.x1),
  y0: Math.min(a.y0, b.y0),
  y1: Math.max(a.y1, b.y1),
});

/** Repeatedly merges near rects; ends with disjoint, well-separated clusters. */
const clustered = (parts: readonly PixelRect[]): PixelRect[] => {
  const clusters = [...parts];
  let changed = true;
  while (changed) {
    changed = false;
    outer: for (let a = 0; a < clusters.length; a += 1) {
      for (let b = a + 1; b < clusters.length; b += 1) {
        if (!near(clusters[a], clusters[b])) continue;
        clusters[a] = merge(clusters[a], clusters[b]);
        clusters.splice(b, 1);
        changed = true;
        break outer;
      }
    }
  }
  return clusters;
};

/**
 * The objects INSIDE the measurement whose centring is worth reporting: the
 * icon, the label, the badge. Detector components sitting within
 * [`CLUSTER_GAP`] of each other are one visual object - a multi-stroke icon
 * must not read as its strokes. The component (or cluster) matching the
 * measurement itself is excluded - comparing a box against itself is always
 * "centred" and says nothing - as are specks.
 */
const innerObjects = (
  boxes: readonly RulerComponentBox[],
  rect: PixelRect,
): readonly PixelRect[] => {
  const inside: { area: number; box: PixelRect }[] = [];
  for (const box of boxes) {
    const candidate = boxRect(box);
    if (!contains(rect, candidate) || isSelf(rect, candidate)) continue;
    const width = candidate.x1 - candidate.x0;
    const height = candidate.y1 - candidate.y0;
    if (width < MINIMUM_OBJECT_SIZE || height < MINIMUM_OBJECT_SIZE) continue;
    inside.push({ area: width * height, box: candidate });
  }
  inside.sort((a, b) => b.area - a.area);
  // Clustering also swallows nested parts: containment implies nearness.
  const clusters = clustered(
    inside.slice(0, MAXIMUM_PARTS).map(({ box }) => box),
  );
  return clusters
    .filter((cluster) => !isSelf(rect, cluster))
    .sort(
      (a, b) => (b.x1 - b.x0) * (b.y1 - b.y0) - (a.x1 - a.x0) * (a.y1 - a.y0),
    )
    .slice(0, MAXIMUM_OBJECTS);
};

const unionOf = (rects: readonly PixelRect[]) =>
  rects.reduce<PixelRect | undefined>(
    (union, rect) =>
      union === undefined
        ? rect
        : {
            x0: Math.min(union.x0, rect.x0),
            x1: Math.max(union.x1, rect.x1),
            y0: Math.min(union.y0, rect.y0),
            y1: Math.max(union.y1, rect.y1),
          },
    undefined,
  );

const rectCentre = (rect: PixelRect, axis: Axis) =>
  axis === "x" ? (rect.x0 + rect.x1) / 2 : (rect.y0 + rect.y1) / 2;

export const centerlines = ({
  bounds,
  boxes,
  deviceScale,
  peers,
}: {
  bounds: Bounds;
  boxes: readonly RulerComponentBox[];
  deviceScale: number;
  peers: readonly Bounds[];
}): Centerlines => {
  const centre = centreOf(bounds);
  const objects = innerObjects(boxes, {
    x0: bounds.x * deviceScale,
    x1: (bounds.x + bounds.width) * deviceScale,
    y0: bounds.y * deviceScale,
    y1: (bounds.y + bounds.height) * deviceScale,
  });
  const union = unionOf(objects);
  const alignedOn = (rect: PixelRect, axis: Axis) => {
    const position = axis === "x" ? centre.x : centre.y;
    const delta = rectCentre(rect, axis) - position * deviceScale;
    return Math.abs(delta) <= ALIGNMENT_SLACK;
  };
  const axisLine = (axis: Axis): Centerline => {
    const position = axis === "x" ? centre.x : centre.y;
    const peerAligned = peers.some((peer) => {
      const other = axis === "x" ? centreOf(peer).x : centreOf(peer).y;
      return Math.abs(other - position) * deviceScale <= ALIGNMENT_SLACK;
    });
    const unionAligned = union !== undefined && alignedOn(union, axis);
    return { accent: peerAligned || unionAligned, position };
  };
  return {
    objects: objects.map((rect) => ({
      alignedX: alignedOn(rect, "x"),
      alignedY: alignedOn(rect, "y"),
      bounds: {
        height: (rect.y1 - rect.y0) / deviceScale,
        width: (rect.x1 - rect.x0) / deviceScale,
        x: rect.x0 / deviceScale,
        y: rect.y0 / deviceScale,
      },
    })),
    x: axisLine("x"),
    y: axisLine("y"),
  };
};

File: src-tauri/src/ruler/centerlines.rs
// SPDX-FileCopyrightText: 2026 overpolish
// SPDX-License-Identifier: GPL-3.0-or-later

use crate::osc::geometry::{Point, Rect};

const SELF_SLACK: f64 = 3.0;
const MINIMUM_OBJECT_SIZE: f64 = 3.0;
const CLUSTER_GAP: f64 = 6.0;
const MAXIMUM_PARTS: usize = 128;
const MAXIMUM_OBJECTS: usize = 12;

#[derive(Clone, Copy, Debug, PartialEq)]
pub(crate) struct InnerObject {
  pub bounds: Rect,
  pub aligned_x: bool,
  pub aligned_y: bool,
}

#[derive(Clone, Debug, PartialEq)]
pub(crate) struct CenterlineAnalysis {
  pub objects: Vec<InnerObject>,
  pub x_accent: bool,
  pub y_accent: bool,
}

fn center(bounds: Rect) -> Point {
  Point {
    x: bounds.origin.x + bounds.size.width * 0.5,
    y: bounds.origin.y + bounds.size.height * 0.5,
  }
}

fn shares_rendered_center(left: f64, right: f64, scale: f64) -> bool {
  (left * scale).round() == (right * scale).round()
}

fn contains(outer: Rect, inner: Rect) -> bool {
  inner.origin.x >= outer.origin.x
    && inner.origin.y >= outer.origin.y
    && inner.right() <= outer.right()
    && inner.bottom() <= outer.bottom()
}

fn is_self(outer: Rect, inner: Rect, slack: f64) -> bool {
  (outer.origin.x - inner.origin.x).abs() <= slack
    && (outer.origin.y - inner.origin.y).abs() <= slack
    && (outer.right() - inner.right()).abs() <= slack
    && (outer.bottom() - inner.bottom()).abs() <= slack
}

fn near(left: Rect, right: Rect, gap: f64) -> bool {
  left.origin.x - gap <= right.right()
    && right.origin.x - gap <= left.right()
    && left.origin.y - gap <= right.bottom()
    && right.origin.y - gap <= left.bottom()
}

fn union(left: Rect, right: Rect) -> Rect {
  let x = left.origin.x.min(right.origin.x);
  let y = left.origin.y.min(right.origin.y);
  Rect::from_xywh(
    x,
    y,
    left.right().max(right.right()) - x,
    left.bottom().max(right.bottom()) - y,
  )
}

fn clustered(mut parts: Vec<Rect>, gap: f64) -> Vec<Rect> {
  let mut changed = true;
  while changed {
    changed = false;
    'outer: for left in 0..parts.len() {
      for right in left + 1..parts.len() {
        if near(parts[left], parts[right], gap) {
          parts[left] = union(parts[left], parts[right]);
          parts.remove(right);
          changed = true;
          break 'outer;
        }
      }
    }
  }
  parts
}

pub(crate) fn analyze(
  bounds: Rect,
  boxes: &[Rect],
  peers: &[Rect],
  device_scale: f64,
) -> CenterlineAnalysis {
  let scale = device_scale.max(f64::EPSILON);
  let self_slack = SELF_SLACK / scale;
  let minimum = MINIMUM_OBJECT_SIZE / scale;
  let mut parts = boxes
    .iter()
    .copied()
    .filter(|candidate| {
      contains(bounds, *candidate)
        && !is_self(bounds, *candidate, self_slack)
        && candidate.size.width >= minimum
        && candidate.size.height >= minimum
    })
    .collect::<Vec<_>>();
  parts.sort_by(|left, right| {
    let left_area = left.size.width * left.size.height;
    let right_area = right.size.width * right.size.height;
    right_area.total_cmp(&left_area)
  });
  parts.truncate(MAXIMUM_PARTS);
  let mut objects = clustered(parts, CLUSTER_GAP / scale)
    .into_iter()
    .filter(|candidate| !is_self(bounds, *candidate, self_slack))
    .collect::<Vec<_>>();
  objects.sort_by(|left, right| {
    let left_area = left.size.width * left.size.height;
    let right_area = right.size.width * right.size.height;
    right_area.total_cmp(&left_area)
  });
  objects.truncate(MAXIMUM_OBJECTS);

  let outer_center = center(bounds);
  let union = objects.iter().copied().reduce(union);
  let x_accent = peers
    .iter()
    .any(|peer| shares_rendered_center(center(*peer).x, outer_center.x, scale))
    || union.is_some_and(|inner| shares_rendered_center(center(inner).x, outer_center.x, scale));
  let y_accent = peers
    .iter()
    .any(|peer| shares_rendered_center(center(*peer).y, outer_center.y, scale))
    || union.is_some_and(|inner| shares_rendered_center(center(inner).y, outer_center.y, scale));
  CenterlineAnalysis {
    objects: objects
      .into_iter()
      .map(|inner| {
        let inner_center = center(inner);
        InnerObject {
          bounds: inner,
          aligned_x: shares_rendered_center(inner_center.x, outer_center.x, scale),
          aligned_y: shares_rendered_center(inner_center.y, outer_center.y, scale),
        }
      })
      .collect(),
    x_accent,
    y_accent,
  }
}

#[cfg(test)]
mod tests {
  use super::*;

  #[test]
  fn clusters_inner_parts_and_reports_content_alignment() {
    let bounds = Rect::from_xywh(10.0, 10.0, 80.0, 60.0);
    let result = analyze(
      bounds,
      &[
        bounds,
        Rect::from_xywh(35.0, 25.0, 12.0, 30.0),
        Rect::from_xywh(52.0, 25.0, 13.0, 30.0),
      ],
      &[],
      1.0,
    );
    assert_eq!(result.objects.len(), 1);
    assert!(result.x_accent);
    assert!(result.y_accent);
    assert!(result.objects[0].aligned_x);
    assert!(result.objects[0].aligned_y);
  }

  #[test]
  fn sibling_centres_accent_only_the_shared_axis() {
    let bounds = Rect::from_xywh(10.0, 10.0, 40.0, 40.0);
    let result = analyze(bounds, &[], &[Rect::from_xywh(10.0, 80.0, 40.0, 20.0)], 2.0);
    assert!(result.x_accent);
    assert!(!result.y_accent);
  }

  #[test]
  fn a_one_device_pixel_offset_is_not_reported_as_centered() {
    let bounds = Rect::from_xywh(10.0, 10.0, 40.0, 40.0);
    let result = analyze(bounds, &[Rect::from_xywh(11.0, 15.0, 39.0, 30.0)], &[], 2.0);
    assert!(!result.x_accent);
    assert!(result.y_accent);
    assert!(!result.objects[0].aligned_x);
    assert!(result.objects[0].aligned_y);
  }
}
